File: modules/core/arrow-kindedj/src/main/java/arrow/kindedj/Convert.java
package arrow.kindedj;

import org.jetbrains.annotations.NotNull;

import arrow.Kind;
import arrow.KindType;

public class Convert extends ForConvert {
    @NotNull
    public static <F, A> FromKindedJToArrow<F, A> fromKindedJ(@NotNull io.kindedj.Hk<F, A> hk) {
        return new FromKindedJToArrow<>(hk);
    }

    @NotNull
    public static <F extends KindType, A> FromArrowToKindedJ<F, A> toKindedJ(@NotNull Kind<F, A> hk) {
        return new FromArrowToKindedJ<>(hk);
    }

    public static class FromKindedJToArrow<F, A> extends Kind<Kind<ForConvert, F>, A> {

        @NotNull
        private final io.kindedj.Hk<F, A> hk;

        FromKindedJToArrow(@NotNull io.kindedj.Hk<F, A> hk) {
            this.hk = hk;
        }

        @NotNull
        public Kind<Kind<ForConvert, F>, A> toArrow() {
            return this;
        }

        @NotNull
        public io.kindedj.Hk<F, A> toKindedJ() {
            return hk;
        }
    }

    public static class FromArrowToKindedJ<F extends KindType, A> implements io.kindedj.Hk<io.kindedj.Hk<ForConvert, F>, A> {

        @NotNull
        private final Kind<F, A> hk;

        FromArrowToKindedJ(@NotNull Kind<F, A> hk) {
            this.hk = hk;
        }

        @NotNull
        public Kind<F, A> toArrow() {
            return hk;
        }

        @NotNull
        public io.kindedj.Hk<io.kindedj.Hk<ForConvert, F>, A> toKindedJ() {
            return this;
        }
    }
}

File: modules/core/arrow-kindedj/src/main/kotlin/arrow/kindedj/Convert.kt
package arrow.kindedj

import arrow.Kind
import arrow.Kind2
import arrow.KindType
import arrow.Witness
import io.kindedj.HkJ2
import io.kindedj.Hk as HK_J

abstract class ForConvert : Witness<Convert>()

@Suppress("UNCHECKED_CAST")
fun <F, A> Kind2<ForConvert, F, A>.fromArrow(): HK_J<F, A> = (this as Convert.FromKindedJToArrow<F, A>).toKindedJ()

fun <F: KindType, A> HkJ2<ForConvert, F, A>.toArrow(): Kind<F, A> = (this as Convert.FromArrowToKindedJ<F, A>).toArrow()

fun <F, A> HK_J<F, A>.fromKindedJ(): Kind2<ForConvert, F, A> = Convert.fromKindedJ(this)

fun <F: KindType, A> Kind<F, A>.toKindedJ(): HkJ2<ForConvert, F, A> = Convert.toKindedJ(this)
